service/controller/v18/resource/nodeindexstatus/create.go

```diff
@@ -0,0 +1,130 @@
+package nodeindexstatus
+
+import (
+	"context"
+	"reflect"
+	"sort"
+
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/operatorkit/controller/context/reconciliationcanceledcontext"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/giantswarm/kvm-operator/service/controller/v18/key"
+)
+
+func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+	cr, err := key.ToCustomObject(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	var nodeIndexes map[string]int
+	{
+		var idx int
+		allocations := key.AllocatedNodeIndexes(cr)
+		nodes := key.AllNodes(cr)
+		nodeIndexes = cr.Status.KVM.NodeIndexes
+		if nodeIndexes == nil {
+			nodeIndexes = make(map[string]int)
+		}
+
+		// Remove node indexes for non-existent nodes.
+		for nodeID, _ := range nodeIndexes {
+			found := false
+
+			for _, n := range nodes {
+				if n.ID == nodeID {
+					found = true
+					break
+				}
+			}
+
+			if !found {
+				delete(nodeIndexes, nodeID)
+			}
+		}
+
+		// Ensure all present nodes have node index allocation.
+		for _, n := range nodes {
+			_, exists := nodeIndexes[n.ID]
+			if !exists {
+				allocations, idx = allocateIndex(allocations)
+				nodeIndexes[n.ID] = idx
+			}
+		}
+	}
+
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", "updating status with node indexes")
+
+		newObj, err := r.g8sClient.ProviderV1alpha1().KVMConfigs(cr.GetNamespace()).Get(cr.GetName(), metav1.GetOptions{})
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		if !reflect.DeepEqual(newObj.Status.KVM.NodeIndexes, nodeIndexes) {
+			newObj.Status.KVM.NodeIndexes = nodeIndexes
+			_, err = r.g8sClient.ProviderV1alpha1().KVMConfigs(newObj.GetNamespace()).UpdateStatus(newObj)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			r.logger.LogCtx(ctx, "level", "debug", "message", "updated status with node indexes")
+
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling reconciliation")
+			reconciliationcanceledcontext.SetCanceled(ctx)
+
+			return nil
+		} else {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "did not update status with node indexes")
+		}
+	}
+
+	{
+		cr.Status.KVM.NodeIndexes = nodeIndexes
+	}
```

Comment:
Why are we doing this here? We assign the indices above. This here looks wrong.

Reply:
This was from one of the `aws-operator` status resources. I was wondering the same that this is wrong, but kept it as it didn't do any harm either. I'll get rid of it.

Reply:
Where did you find it that way?

Reply:
Most probably it was this:
https://github.com/giantswarm/aws-operator/blob/master/service/controller/v22/resource/asgstatus/create.go#L108-L112

Reply:
What we do there is setting the controller context. Here you set the CR status.

Reply:
Oh wow! You're right. I confused `cr/cc/newObj` there and thought that it was somehow necessary. Oh well, I removed that status part in my code so that should be correct now.